File: packages/ts/src/exceptions.ts
/** Base authentication/authorization error. */
export class AuthError extends Error {
  public readonly statusCode: number;
  public readonly detail: string;

  constructor(detail = "Authentication error", statusCode = 401) {
    super(detail);
    this.name = "AuthError";
    this.detail = detail;
    this.statusCode = statusCode;
  }
}

export class InvalidTokenError extends AuthError {
  constructor(detail = "Invalid token") {
    super(detail, 401);
    this.name = "InvalidTokenError";
  }
}

export class TokenExpiredError extends AuthError {
  constructor(detail = "Token has expired") {
    super(detail, 401);
    this.name = "TokenExpiredError";
  }
}

export class UnauthorizedError extends AuthError {
  constructor(detail = "Not authenticated") {
    super(detail, 401);
    this.name = "UnauthorizedError";
  }
}

export class ForbiddenError extends AuthError {
  constructor(detail = "Forbidden") {
    super(detail, 403);
    this.name = "ForbiddenError";
  }
}

File: packages/py/src/urauth/exceptions.py
"""Core exceptions — plain Python, no framework dependency."""

from __future__ import annotations


class AuthError(Exception):
    """Base authentication/authorization error."""

    def __init__(self, detail: str = "Authentication error", *, status_code: int = 401) -> None:
        self.detail = detail
        self.status_code = status_code
        super().__init__(detail)


class InvalidTokenError(AuthError):
    """Token is malformed or signature verification failed."""

    def __init__(self, detail: str = "Invalid token") -> None:
        super().__init__(detail, status_code=401)


class TokenExpiredError(AuthError):
    """Token has expired."""

    def __init__(self, detail: str = "Token has expired") -> None:
        super().__init__(detail, status_code=401)


class TokenRevokedError(AuthError):
    """Token has been revoked."""

    def __init__(self, detail: str = "Token has been revoked") -> None:
        super().__init__(detail, status_code=401)


class UnauthorizedError(AuthError):
    """User is not authenticated."""

    def __init__(self, detail: str = "Not authenticated") -> None:
        super().__init__(detail, status_code=401)


class ForbiddenError(AuthError):
    """User lacks required permissions/roles."""

    def __init__(self, detail: str = "Forbidden") -> None:
        super().__init__(detail, status_code=403)
